File: 임현규/programmers/Lv2/귤고르기/python/귤고르기.py
from collections import Counter


def solution(k, tangerine) -> int:
    counter: Counter = Counter(tangerine)
    common: list[tuple[int, int]] = counter.most_common()
    result: int = 0
    idx: int = 0
    while k > 0 and idx < len(common):
        value: int = common[idx][1]
        k -= value
        idx += 1
        result += 1
    return result



result = solution(2, [1, 1, 1, 1, 2, 2, 2, 3])
print(result)

File: 임현규/programmers/Lv2/귤고르기/java/귤고르기.java
import java.util.Arrays;
import java.util.Comparator;
import java.util.Map;
import java.util.Objects;
import java.util.PriorityQueue;
import java.util.function.Function;
import java.util.stream.Collectors;

public class Solution {

    public int solution(int k, int[] tangerine) {
        PriorityQueue<Long> pq = makeFrequencyCountMaxHeap(tangerine);
        int result = 0;
        while (k > 0) {
            int count = Objects.requireNonNull(pq.poll()).intValue();
            k -= count;
            result += 1;
        }
        return result;
    }

    private Map<Integer, Long> makeFrequency(int[] tangerine) {
        return Arrays.stream(tangerine).boxed()
            .collect(Collectors.groupingBy(Function.identity(), Collectors.counting()));
    }

    private PriorityQueue<Long> makeFrequencyCountMaxHeap(int[] tangerine) {
        PriorityQueue<Long> pq = new PriorityQueue<>(
            Comparator.comparing(value -> value, Comparator.reverseOrder()));
        pq.addAll(makeFrequency(tangerine).values());
        return pq;
    }
}